bgpd/bgp_orr.c

```diff
@@ -0,0 +1,1163 @@
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "bgpd/bgpd.h"
+#include "bgpd/bgp_debug.h"
+#include "bgpd/bgp_orr.h"
+#include "bgpd/bgp_vty.h"
+#include "zclient.h"
+
+DEFINE_MTYPE_STATIC(BGPD, ORR_IGP_INFO, "ORR IGP Metric info");
+
+extern struct zclient *zclient;
+
+static inline bool is_orr_primary_root(struct bgp_orr_group *orr_group,
+				       char *host)
+{
+	return orr_group->primary && !strcmp(orr_group->primary->host, host);
+}
+
+static inline bool is_orr_secondary_root(struct bgp_orr_group *orr_group,
+					 char *host)
+{
+	return orr_group->secondary &&
+	       !strcmp(orr_group->secondary->host, host);
+}
+
+static inline bool is_orr_tertiary_root(struct bgp_orr_group *orr_group,
+					char *host)
+{
+	return orr_group->tertiary && !strcmp(orr_group->tertiary->host, host);
+}
+
+static inline bool is_orr_active_root(struct bgp_orr_group *orr_group,
+				      char *host)
+{
+	return orr_group->active && !strcmp(orr_group->active->host, host);
+}
+static inline bool is_orr_root_node(struct bgp_orr_group *orr_group, char *host)
+{
+	return is_orr_primary_root(orr_group, host) ||
+	       is_orr_secondary_root(orr_group, host) ||
+	       is_orr_tertiary_root(orr_group, host);
+}
+
+static inline bool is_orr_primary_reachable(struct bgp_orr_group *orr_group)
+{
+	return orr_group->primary &&
+	       orr_group->primary->afc_nego[orr_group->afi][orr_group->safi] &&
+	       orr_group->primary->status == Established;
+}
+
+static inline bool is_orr_secondary_reachable(struct bgp_orr_group *orr_group)
+{
+	return orr_group->secondary &&
+	       orr_group->secondary
+		       ->afc_nego[orr_group->afi][orr_group->safi] &&
+	       orr_group->secondary->status == Established;
+}
+
+static inline bool is_orr_tertiary_reachable(struct bgp_orr_group *orr_group)
+{
+	return orr_group->tertiary &&
+	       orr_group->tertiary->afc_nego[orr_group->afi][orr_group->safi] &&
+	       orr_group->tertiary->status == Established;
+}
+
+static inline bool is_peer_orr_group_member(struct peer *peer, afi_t afi,
+					    safi_t safi, const char *name)
+{
+	return peer_af_flag_check(peer, afi, safi, PEER_FLAG_ORR_GROUP) &&
+	       !strcmp(peer->orr_group_name[afi][safi], name);
+}
+
+static inline bool is_peer_reachable(struct peer *peer, afi_t afi, safi_t safi)
```

Comment:
isn't there some other already-existing function for this?

Reply:
@riw777 I checked, but looks like there is no existing common function to check the afc_nego[][] and status. they are individually checked as and when needed.
@ton31337 could you please point me to the function if we have any such function?

Reply:
AFAIK, we don't have it. I suggest just reusing peer_established() also for the last test.